cachix/src/Cachix/Client/Push.hs

```diff
@@ -76,102 +81,104 @@ pushSingleStorePath
+      name = pushCacheName cache
+
+  onAttempt cb retrystatus
+
+  narSizeRef <- liftIO $ newIORef 0
+  fileSizeRef <- liftIO $ newIORef 0
+  narHashRef <- liftIO $ newIORef ("" :: ByteString)
+  fileHashRef <- liftIO $ newIORef ("" :: ByteString)
+
+  -- stream store path as xz compressed nar file
+  nar <- liftIO $ runPut . Binary.put <$> Nar.localPackNar Nar.narEffectsIO (toS storePath)
+
+  withXzipCompressor cb $ \xzCompressor -> do
+    let stream'
+          = sourceLazy nar
+          .| passthroughSizeSink narSizeRef
+          .| passthroughHashSink narHashRef
+          .| xzCompressor
+          .| passthroughSizeSink fileSizeRef
+          .| passthroughHashSinkB16 fileHashRef
+
+    -- for now we need to use letsencrypt domain instead of cloudflare due to its upload limits
+    let newClientEnv = clientEnv {
+            baseUrl = (baseUrl clientEnv) { baseUrlHost = toS name <> "." <> baseUrlHost (baseUrl clientEnv)}
+          }
+    (_ :: NoContent) <- liftIO $ (`withClientM` newClientEnv)
+        (Api.createNar (cachixBCStreamingClient name) stream')
+        $ escalate >=> \NoContent -> do
+            -- TODO: we're done, so we can leave withClientM. Doing so
+            --       will allow more concurrent requests when the number
+            --       of XZ compressors is limited
+            narSize <- readIORef narSizeRef
+            narHash <- ("sha256:" <>) . System.Nix.Base32.encode <$> readIORef narHashRef
+            fileHash <- readIORef fileHashRef
+            fileSize <- readIORef fileSizeRef
+
+            deriverRaw <- T.strip . toS <$> readProcess "nix-store" ["-q", "--deriver", toS storePath] mempty
+            let deriver = if deriverRaw == "unknown-deriver"
+                          then deriverRaw
+                          else T.drop 11 deriverRaw
+
+            references <- sort . T.lines . T.strip . toS <$> readProcess "nix-store" ["-q", "--references", toS storePath] mempty
+
+            let
+                fp = fingerprint storePath narHash narSize references
+                sig = dsign (signingSecretKey $ pushCacheSigningKey cache) fp
+                nic = Api.NarInfoCreate
+                  { cStoreHash = storeHash
+                  , cStoreSuffix = storeSuffix
+                  , cNarHash = narHash
+                  , cNarSize = narSize
+                  , cFileSize = fileSize
+                  , cFileHash = toS fileHash
+                  , cReferences = fmap (T.drop 11) references
+                  , cDeriver = deriver
+                  , cSig = toS $ B64.encode $ unSignature sig
+                  }
+
+            escalate $ Api.isNarInfoCreateValid nic
+
+            -- Upload narinfo with signature
+            escalate <=< (`runClientM` clientEnv) $ Api.createNarinfo
+              (cachixBCClient name)
+              (Api.NarInfoC storeHash)
+              nic
+    onDone cb
+
+-- Retry up to 5 times for each store path.
```

Comment:
Why is it 5 instead of 3? Do we need to specify it in this comment?
It should probably be a haddock.
This function is not only used for paths, so the name should change.